cmd/frontend/graphqlbackend/repository_comparison.go

```diff
@@ -348,12 +351,102 @@ func diffPathOrNull(path string) *string {
 	return &path
 }
 
-func NewDiffHunk(hunk *diff.Hunk) *DiffHunk {
-	return &DiffHunk{hunk: hunk}
+func NewDiffHunk(hunk *diff.Hunk, highlighter DiffHighlighter) *DiffHunk {
+	return &DiffHunk{hunk: hunk, highlighter: highlighter}
+}
+
+type DiffHighlighter interface {
+	Highlight(ctx context.Context, args *HighlightArgs) (map[int32]string, map[int32]string, bool, error)
+}
+
+type HighlightArgs struct {
+	DisableTimeout     bool
+	IsLightTheme       bool
+	HighlightLongLines bool
+}
+
+type fileDiffHighlighter struct {
+	fileDiffResolver *fileDiffResolver
+	highlightedBase  map[int32]string
+	highlightedHead  map[int32]string
+	highlightOnce    sync.Once
+	highlightErr     error
+	highlightAborted bool
+}
+
+func (r *fileDiffHighlighter) Highlight(ctx context.Context, args *HighlightArgs) (map[int32]string, map[int32]string, bool, error) {
+	r.highlightOnce.Do(func() {
+		if oldFile := r.fileDiffResolver.OldFile(); oldFile != nil {
+			var binary bool
+			binary, r.highlightErr = oldFile.Binary(ctx)
+			if r.highlightErr != nil {
+				return
+			}
+			if !binary {
```

Comment:
Go style would be to do an early exit here
```suggestion
			if binary {
				return
			}
```